File: pipeline/web/static/js/claimreview.js
/* The Claim review tab: the human-review labour BETA-034 is blocked on (BETA-047).
 *
 * A reviewer works one machine-extracted (subject, predicate, object) triple
 * at a time and records approve / reject / corrected against their name. The
 * backend functions already exist -- `pipeline/nlp/decisions.decide` (with
 * ontology-validated corrections), `decisions.history`, and
 * `pipeline/nlp/gate.check` for the training-readiness report. This tab is
 * the workbench over them, and it adds no policy of its own:
 *
 *   * one candidate per decision -- there is no bulk-approve control here;
 *   * a decision without a reviewer name is refused (server-side too);
 *   * nothing writes graph_claims, trains a classifier, or produces public
 *     output. The copy says so and the code has no path to it.
 *
 * An /api/admin/* tool, behind the operator's network-trust boundary.
 */
import { el } from './dom.js';

const $ = (id) => document.getElementById(id);

const PAGE = 25;
const state = {
  offset: 0,
  predicates: [],
  concepts: [],
  reasonCodes: [],
};

async function api(path, options) {
  const response = await fetch(path, options);
  let payload = null;
  try { payload = await response.json(); } catch (e) { /* not JSON */ }
  if (!response.ok) {
    const error = new Error((payload && payload.error) || response.statusText);
    error.status = response.status;
    throw error;
  }
  return payload;
}

function reviewerName() {
  const field = $('reviewer');
  return field ? field.value.trim() : '';
}

function filters() {
  return {
    q: $('cr-q').value.trim(),
    status: $('cr-status').value,
    predicate: $('cr-predicate').value,
    source_system: $('cr-source').value,
  };
}

// --- the gate strip ---------------------------------------------------------

async function loadGate() {
  const box = $('claimreview-gate');
  let data;
  try { data = await api('/api/admin/claim-gate'); }
  catch (e) { box.hidden = true; return; }

  const chips = [
    el('span', { class: 'chip', text: data.ready ? 'gate: READY' : 'gate: not yet' }),
    el('span', { class: 'chip', text: `${data.n_decisions} decisions` }),
    el('span', { class: 'chip', text: `${data.n_decided_candidates} candidates decided` }),
  ];
  const inter = data.inter_reviewer || {};
  chips.push(el('span', { class: 'chip',
    text: `double-reviewed ${inter.double_reviewed ?? 0}${inter.assessed ? `, agreement ${inter.agreement}` : ' (not assessed)'}` }));

  const catRows = Object.entries(data.categories || {}).map(([name, c]) =>
    el('li', {},
      el('strong', { text: name }),
      ` — +${c.positive} / -${c.negative}, ${c.distinct_subjects} subjects, `
      + `${(c.years || []).length} years — `,
      el('span', { text: c.ready ? 'ready' : (c.shortfalls || []).join('; ') })));

  box.replaceChildren(
    el('div', { class: 'chips' }, ...chips),
    el('p', { class: 'small muted', style: 'margin:6px 0 2px',
      text: 'Per-category training readiness (034G). Closing the gate is reviewer labour, not code:' }),
    el('ul', { class: 'small' }, ...catRows),
    (data.blocking || []).length
      ? el('p', { class: 'small', text: `Blocking: ${data.blocking.join(' · ')}` })
      : null);
  box.hidden = false;
}

// --- the vocabularies for the correction form ------------------------------

async function loadOntology() {
  let data;
  try { data = await api('/api/admin/claim-ontology'); }
  catch (e) { return; }
  state.predicates = data.predicates || [];
  state.concepts = data.concepts || [];
  state.reasonCodes = data.reason_codes || [];

  const sel = $('cr-predicate');
  sel.replaceChildren(el('option', { value: '', text: 'Any predicate' }));
  for (const p of state.predicates) {
    sel.append(el('option', { value: p.id, text: `${p.label} (${p.id})` }));
  }
}

// --- one candidate + its decision form ------------------------------------

function decisionForm(candidate) {
  const wrap = el('div', { class: 'panel' });
  const status = el('div', { class: 'small' });

  const decision = el('select', { 'aria-label': 'Decision' },
    el('option', { value: 'approved', text: 'Approve — triple is right' }),
    el('option', { value: 'rejected', text: 'Reject — wrong, not salvageable' }),
    el('option', { value: 'corrected', text: 'Correct — real, but a field is wrong' }));

  const reason = el('select', { 'aria-label': 'Reason code' },
    el('option', { value: '', text: '(reason code, optional)' }),
    ...state.reasonCodes.map((r) => el('option', { value: r, text: r })));

  const correctedPredicate = el('select', { 'aria-label': 'Corrected predicate' },
    el('option', { value: '', text: '(keep predicate)' }),
    ...state.predicates.map((p) => el('option', { value: p.id, text: p.label })));
  const correctedConcept = el('select', { 'aria-label': 'Corrected object concept' },
    el('option', { value: '', text: '(keep object concept)' }),
    ...state.concepts.map((c) => el('option', { value: c.id, text: `${c.label} (${c.id})` })));
  const correctedLiteral = el('input', { type: 'text', 'aria-label': 'Corrected object literal', placeholder: 'corrected object literal' });

  const correctionRow = el('div', { class: 'runbar', hidden: true },
    correctedPredicate, correctedConcept, correctedLiteral);
  decision.addEventListener('change', () => {
    correctionRow.hidden = decision.value !== 'corrected';
  });

  const note = el('input', { type: 'text', 'aria-label': 'Note', placeholder: 'note (optional)' });
  const submit = el('button', { class: 'btn primary', type: 'submit', text: 'Record decision' });

  const form = el('form', { class: 'stack' },
    el('div', { class: 'runbar' }, decision, reason),
    correctionRow,
    el('div', { class: 'runbar' }, note, submit),
    status);

  form.addEventListener('submit', async (event) => {
    event.preventDefault();
    const who = reviewerName();
    if (!who) { status.textContent = 'Enter your name in the Reviewer box above.'; return; }
    submit.disabled = true;
    status.textContent = 'Recording…';
    const body = {
      claim_candidate_id: candidate.claim_candidate_id,
      decision: decision.value,
      decided_by: who,
      reason_code: reason.value || null,
      note: note.value.trim() || null,
    };
    if (decision.value === 'corrected') {
      body.corrected_predicate = correctedPredicate.value || null;
      body.corrected_object_concept_id = correctedConcept.value || null;
      body.corrected_object_literal = correctedLiteral.value.trim() || null;
    }
    try {
      const result = await api('/api/admin/claim-candidates/decide', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });
      status.textContent = `Recorded: ${result.decision} → ${result.status}.`;
      loadGate();
      loadList();
    } catch (error) {
      status.textContent = `Refused: ${error.message}`;
      submit.disabled = false;
    }
  });

  wrap.replaceChildren(form);
  return wrap;
}

function candidateCard(candidate) {
  const objectText = candidate.object_concept_label
    ? `${candidate.object_concept_label} (${candidate.object_concept_id})`
    : (candidate.object_literal || '—');
  const meta = [candidate.source_system, candidate.document_type,
    `score ${candidate.relation_score}`, candidate.assertion_status,
    `status ${candidate.status}`].filter(Boolean).join(' · ');

  return el('article', { class: 'panel' },
    el('div', { class: 'row', style: 'justify-content:space-between;gap:8px;align-items:baseline' },
      el('strong', { text: candidate.predicate_label || candidate.predicate }),
      el('span', { class: 'muted small', text: meta })),
    el('p', { class: 'small' },
      el('em', { text: `subject: ${candidate.subject_hint || '(from mention)'}` }),
      ` — object: ${objectText}`),
    el('blockquote', { class: 'small', text: candidate.evidence_span || '' }),
    el('div', { class: 'row small', style: 'gap:12px;flex-wrap:wrap' },
      candidate.source_url
        ? el('a', { href: candidate.source_url, target: '_blank', rel: 'noopener' }, 'Source page')
        : null,
      candidate.last_decision
        ? el('span', { class: 'muted', text: `last: ${candidate.last_decision} by ${candidate.last_decided_by} (${candidate.decision_count})` })
        : el('span', { class: 'muted', text: 'no decision yet' })),
    decisionForm(candidate));
}

async function loadList() {
  const list = $('claimreview-list');
  const statusLine = $('claimreview-status');
  const f = filters();
  const params = new URLSearchParams({ offset: state.offset, limit: PAGE });
  if (f.q) params.set('q', f.q);
  if (f.status) params.set('status', f.status);
  if (f.predicate) params.set('predicate', f.predicate);
  if (f.source_system) params.set('source_system', f.source_system);

  statusLine.textContent = 'Loading…';
  let data;
  try { data = await api(`/api/admin/claim-candidates?${params}`); }
  catch (error) { statusLine.textContent = `Failed: ${error.message}`; return; }

  const rows = data.candidates || [];
  statusLine.textContent = data.total
    ? `${data.page.offset + 1}–${data.page.offset + rows.length} of ${data.total} candidates.`
    : 'No candidates match these filters.';
  list.replaceChildren(...rows.map(candidateCard));

  const pager = $('claimreview-pager');
  pager.replaceChildren();
  if (state.offset > 0) {
    pager.append(el('button', { class: 'btn', type: 'button',
      onclick: () => { state.offset = Math.max(0, state.offset - PAGE); loadList(); } }, '← Previous'));
  }
  if (data.page.offset + rows.length < data.total) {
    pager.append(el('button', { class: 'btn', type: 'button',
      onclick: () => { state.offset += PAGE; loadList(); } }, 'Next →'));
  }
}

export function initClaimReview() {
  const panel = $('tab-claimreview');
  if (!panel) return;

  $('claimreview-filters').addEventListener('submit', (event) => {
    event.preventDefault();
    state.offset = 0;
    loadList();
  });

  let loaded = false;
  const load = () => {
    if (loaded) return;
    loaded = true;
    loadOntology().then(loadList);
    loadGate();
  };
  const observer = new MutationObserver(() => {
    if (panel.classList.contains('active')) load();
  });
  observer.observe(panel, { attributes: true, attributeFilter: ['class'] });
  if (panel.classList.contains('active')) load();
}

File: pipeline/web/claim_review.py
"""Operator-side workbench for the semantic claim-candidate review (BETA-047).

`pipeline/nlp/` already has the parts a reviewer needs — `decisions.decide`
records one person's judgement (and, for `corrected`, their better predicate /
object / subject, ontology-validated), `decisions.history` is the audit trail,
and `gate.check` reports exactly how far the review corpus is from letting the
034G classifiers be trained. What was missing is a place to *do* the review.

This is the one-way bridge, the same shape as `pipeline/web/semantic.py`: it
turns query-strings and JSON bodies into calls on those functions and their
`ClaimDecisionError` into the `QueryError` the request handler already turns
into a 400. It adds no policy of its own:

  * decisions are recorded one candidate at a time, never a list — there is
    no bulk-approve path here;
  * nothing writes `graph_claims` (that first-writer decision is parked, see
    beta.md's BETA-034 `context_034f_graph`);
  * nothing trains a model or produces public output.

An `/api/admin/*` tool: it reads the parsed archive, holds no `restricted_`
data, and stays behind the operator's network-trust boundary.
"""
from __future__ import annotations

from pipeline.nlp import decisions
from pipeline.nlp import gate as gate_mod
from pipeline.nlp import ontology as ontology_mod
from pipeline.web.queries import QueryError

PAGE = 25

# `document_claim_candidates.status`: 'new' (extracted, not yet queued),
# 'queued' (a review_queue item exists), 'accepted' / 'dismissed' (a person
# decided). A reviewer filters on these.
STATUSES = ("new", "queued", "accepted", "dismissed")

_BASE_JOIN = """
    FROM document_claim_candidates c
    JOIN document_chunks dc ON dc.document_chunk_id = c.document_chunk_id
    JOIN document_versions v ON v.document_version_id = dc.document_version_id
    JOIN document_records d ON d.document_id = v.document_id
    JOIN evidence_records e ON e.evidence_id = d.evidence_id
"""


def _rows(conn, sql, params=()):
    return [dict(row) for row in conn.execute(sql, params).fetchall()]


def _label_map():
    onto = ontology_mod.default()
    return (
        {rid: rel.label for rid, rel in onto.relations.items()},
        {cid: con.label for cid, con in onto.concepts.items()},
        onto.version,
    )


def _decorate(row, relation_labels, concept_labels):
    row["predicate_label"] = relation_labels.get(row.get("predicate"))
    row["object_concept_label"] = concept_labels.get(row.get("object_concept_id"))
    return row


def listing(conn, *, status: str | None = None, predicate: str | None = None,
            source_system: str | None = None, q: str | None = None,
            offset: int = 0, limit: int = PAGE) -> dict:
    """A page of live claim candidates with their triple, evidence and the
    latest decision on each."""
    if status and status not in STATUSES:
        raise QueryError(f"status must be one of {', '.join(STATUSES)}.")
    limit = max(1, min(int(limit), 200))
    offset = max(0, int(offset))

    where = ["c.superseded = 0"]
    params: list = []
    if status:
        where.append("c.status = %s")
        params.append(status)
    if predicate:
        where.append("c.predicate = %s")
        params.append(predicate)
    if source_system:
        where.append("e.source_system = %s")
        params.append(source_system)
    if q:
        where.append("(c.evidence_span LIKE %s OR c.subject_hint LIKE %s "
                     "OR c.object_literal LIKE %s)")
        params.extend([f"%{q}%"] * 3)
    clause = " AND ".join(where)

    total = conn.execute(
        f"SELECT COUNT(*) AS count {_BASE_JOIN} WHERE {clause}", params).fetchone()["count"]

    rows = _rows(conn, f"""
        SELECT c.claim_candidate_id, c.predicate, c.object_concept_id,
               c.object_literal, c.assertion_status, c.subject_hint,
               c.relation_score, c.relation_extractor, c.relation_extractor_version,
               c.evidence_span, c.status, c.created_at,
               e.source_system, e.source_url, e.retrieved_at,
               d.title AS document_title, d.document_type,
               (SELECT dd.decision FROM claim_candidate_decisions dd
                WHERE dd.claim_candidate_id = c.claim_candidate_id
                ORDER BY dd.decided_at DESC LIMIT 1) AS last_decision,
               (SELECT dd.decided_by FROM claim_candidate_decisions dd
                WHERE dd.claim_candidate_id = c.claim_candidate_id
                ORDER BY dd.decided_at DESC LIMIT 1) AS last_decided_by,
               (SELECT COUNT(*) FROM claim_candidate_decisions dd
                WHERE dd.claim_candidate_id = c.claim_candidate_id) AS decision_count
        {_BASE_JOIN}
        WHERE {clause}
        ORDER BY c.relation_score DESC, c.claim_candidate_id
        LIMIT %s OFFSET %s""", [*params, limit, offset])

    relation_labels, concept_labels, _ = _label_map()
    for row in rows:
        _decorate(row, relation_labels, concept_labels)

    return {
        "candidates": rows,
        "total": total,
        "page": {"offset": offset, "limit": limit, "returned": len(rows),
                 "status": status, "predicate": predicate,
                 "source_system": source_system, "q": q},
        "caveat": (
            "A candidate is a machine-extracted (subject, predicate, object) "
            "triple from one sentence — a lead for a reviewer, not a claim. "
            "The relevance score ranks for review only. Deciding one records "
            "a named judgement; it does not write a graph claim or train a "
            "model."),
    }


def detail(conn, claim_candidate_id: str) -> dict:
    """One candidate: its triple, the sentence and containing chunk, source
    identity, and every decision recorded against it."""
    row = conn.execute(f"""
        SELECT c.*, dc.text AS chunk_text, dc.page_start, dc.page_end,
               e.source_system, e.source_url, e.retrieved_at, e.payload_sha256,
               d.title AS document_title, d.document_type, d.published_at,
               d.document_id
        {_BASE_JOIN}
        WHERE c.claim_candidate_id = %s AND c.superseded = 0
        """, (claim_candidate_id,)).fetchone()
    if row is None:
        raise QueryError(f"No live claim candidate {claim_candidate_id!r}.")
    row = dict(row)

    relation_labels, concept_labels, ontology_version = _label_map()
    _decorate(row, relation_labels, concept_labels)
    row["ontology_version"] = ontology_version
    row["decisions"] = decisions.history(conn, claim_candidate_id)
    return row


def gate(conn) -> dict:
    """The read-only 034G gate report: per-category decided counts, spread and
    the blocking list. `pipeline/nlp/gate.check` verbatim — no mutation."""
    return gate_mod.check(conn)


def ontology_options(conn=None) -> dict:
    """The controlled vocabularies a `corrected` decision must choose from —
    for the correction form's dropdowns, so a reviewer cannot type a
    predicate or concept the ontology does not have."""
    onto = ontology_mod.default()
    return {
        "ontology_version": onto.version,
        "predicates": [
            {"id": rid, "label": rel.label, "object": rel.object}
            for rid, rel in sorted(onto.relations.items())],
        "concepts": [
            {"id": cid, "label": con.label,
             "categories": sorted(con.categories)}
            for cid, con in sorted(onto.concepts.items())],
        "reason_codes": ["wrong_predicate", "wrong_object", "wrong_subject",
                          "not_a_claim", "out_of_scope", "unclear_sentence"],
    }


def decide(conn, *, claim_candidate_id: str, decision: str, decided_by: str,
           reason_code: str | None = None,
           corrected_predicate: str | None = None,
           corrected_object_concept_id: str | None = None,
           corrected_object_literal: str | None = None,
           corrected_subject_mention_id: str | None = None,
           review_queue_id: int | None = None,
           note: str | None = None) -> dict:
    """One candidate, one named reviewer. Wraps `decisions.decide` and turns
    its `ClaimDecisionError` into the handler's 400."""
    try:
        return decisions.decide(
            conn, claim_candidate_id, decision, decided_by,
            reason_code=reason_code,
            corrected_predicate=corrected_predicate,
            corrected_object_concept_id=corrected_object_concept_id,
            corrected_object_literal=corrected_object_literal,
            corrected_subject_mention_id=corrected_subject_mention_id,
            review_queue_id=review_queue_id, note=note)
    except decisions.ClaimDecisionError as exc:
        raise QueryError(str(exc)) from None
